File: controllers/userQuestionsController.js
import initKnex from "knex";
import configuration from "../knexfile.js";
const knex = initKnex(configuration);

const index = async (req, res) => {
  try {
    let questions = await knex.select("*").from("user_questions").where({id: req.id});
    res.status(200).json(questions);
  } catch (error) {
    console.error(`Error retrieving questions:${error}`);
    res.status(400).json({ message: `Error retrieving questions:${error}` });
  }
};
//make sure only unique questions are added
const addOne = async (req, res) => {
  try {
    const {
      structured_question,
      leetcode_id,
      question_name,
      question_slug,
      question_difficulty,
      unstructured_question_body,
      user_id,
    } = req.body;
    let newQuestion;
    if (structured_question) {
      if (
        !leetcode_id ||
        !question_name ||
        !question_slug ||
        !question_difficulty ||
        !user_id
      ) {
        return res.status(400).send("All fields are required");
      }
      //do a get here to check if leetcode_id in user_question if so do not add a new question
      let newQuestionId = await knex("user_questions").insert({
        leetcode_id,
        question_name,
        question_slug,
        question_difficulty,
        structured_question,
      });
      newQuestion = await knex("user_questions").where({
        id: newQuestionId[0],
      });
    } else {
      if (!unstructured_question_body || !question_name || !user_id) {
        return res.status(400).send("All fields are required");
      }
      let newQuestionId = await knex("user_questions").insert({
        question_name,
        unstructured_question_body,
        structured_question,
      });
      newQuestion = await knex("user_questions").where({
        id: newQuestionId[0],
      });
    }
    return res.status(201).json(newQuestion[0]);
  } catch (error) {
    console.error("Error creating question: ", error);
    res.status(400).json({ message: `Unable to create new one: ${error}` });
  }
};

const addSubmission = async (req, res) => {
  try {
    const { id: user_question_id } = req.params;
    const { submission_analyses } = req.body;
    const newSubmissionId = await knex("submissions").insert({
      user_question_id,
      submission_analyses,
    });
    const newSubmission = await knex
      .select("*")
      .from("submissions")
      .where({ id: newSubmissionId[0] });
    res.status(201).json(newSubmission[0]);
  } catch (error) {
    console.error(`Error adding submission:${error}`);
    res.status(400).json({ message: `Error adding submission:${error}` });
  }
};
const getOne = async (req, res) => {
  try {
    const { id } = req.params;
    let question = await knex("user_questions").where({ id: id });
    if (!question) {
      return res.status(404).json({ message: "Question with id not found" });
    }
    const submissions = await knex
      .select("*")
      .from("submissions")
      .where({ user_question_id: id });
    return res.status(200).json({ question: question[0], submissions });
  } catch (error) {
    console.error(`Error retrieving question:${error}`);
    res.status(400).json({ message: `Error retrieving question:${error}` });
  }
};
export { index, addOne, getOne, addSubmission };

File: Controllers/UserQuestionsController.cs
using Microsoft.AspNetCore.Authorization;
using System.Text.Json.Serialization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using CodeCompareServer.Data;
using CodeCompareServer.Models;

namespace CodeCompareServer.Controllers;

[ApiController]
[Route("userQuestions")]
// [Authorize]
public class UserQuestionsController : ControllerBase
{
    private readonly AppDbContext _context;

    public UserQuestionsController(AppDbContext context)
    {
        _context = context;
    }

    private int GetUserId()
    {
        var idClaim = User.FindFirst("id")?.Value;
        return int.TryParse(idClaim, out int id) ? id : throw new UnauthorizedAccessException();
    }

    [HttpGet]
    public async Task<IActionResult> Index()
    {
        try
        {
            var userId = GetUserId();
            var questions = await _context.UserQuestions
                .Where(uq => uq.UserId == userId)
                .ToListAsync();
            return Ok(questions);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error retrieving questions:{ex}");
            return BadRequest(new { message = $"Error retrieving questions:{ex.Message}" });
        }
    }

    public record AddQuestionRequest(
        [property: JsonPropertyName("structured_question")] bool StructuredQuestion,
        [property: JsonPropertyName("question_name")] string? QuestionName,
        [property: JsonPropertyName("question_slug")] string? QuestionSlug,
        [property: JsonPropertyName("question_difficulty")] string? QuestionDifficulty,
        [property: JsonPropertyName("unstructured_question_body")] string? UnstructuredQuestionBody
    );

    [HttpPost]
    public async Task<IActionResult> AddOne([FromBody] AddQuestionRequest req)
    {
        try
        {

            if (req.StructuredQuestion)
            {
                if (string.IsNullOrEmpty(req.QuestionName) || string.IsNullOrEmpty(req.QuestionSlug) || string.IsNullOrEmpty(req.QuestionDifficulty))
                {
                    return BadRequest("All fields are required");
                }
            }
            else
            {
                if (string.IsNullOrEmpty(req.UnstructuredQuestionBody) || string.IsNullOrEmpty(req.QuestionName))
                {
                    return BadRequest("All fields are required");
                }
            }

            var userId = GetUserId();
            UserQuestion newQuestion;

            if (req.StructuredQuestion)
            {
                var allQuestion = await _context.AllQuestions.FirstOrDefaultAsync(q => q.TitleSlug == req.QuestionSlug);
                if (allQuestion == null) return BadRequest(new { message = "could not find question" });

                newQuestion = new UserQuestion
                {
                    QuestionId = allQuestion.Id,
                    QuestionName = req.QuestionName,
                    QuestionSlug = req.QuestionSlug,
                    QuestionDifficulty = req.QuestionDifficulty,
                    StructuredQuestion = req.StructuredQuestion,
                    UserId = userId,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow
                };
            }
            else
            {
                newQuestion = new UserQuestion
                {
                    QuestionName = req.QuestionName,
                    UnstructuredQuestionBody = req.UnstructuredQuestionBody,
                    StructuredQuestion = false,
                    UserId = userId,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow
                };
            }

            _context.UserQuestions.Add(newQuestion);
            await _context.SaveChangesAsync();
            Console.WriteLine($"4 Question saved with id: {newQuestion.Id}");
            return StatusCode(201, newQuestion);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("Error creating question: " + ex);
            return BadRequest(new { message = $"Unable to create new one: {ex.Message}" });
        }
    }
    [HttpGet("{id}")]
    public async Task<IActionResult> GetOne(int id)
    {
        try
        {
            var question = await _context.UserQuestions.FirstOrDefaultAsync(q => q.Id == id);
            if (question == null)
            {
                return NotFound(new { message = "Question with id not found" });
            }

            var submissions = await _context.Submissions
                .Where(s => s.UserQuestionId == id)
                .ToListAsync();

            return Ok(new { question, submissions });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error retrieving question:{ex}");
            return BadRequest(new { message = $"Error retrieving question:{ex.Message}" });
        }
    }

    public record AddSubmissionRequest(
        [property: JsonPropertyName("submission_analyses")] string SubmissionAnalyses
    );

    [HttpPost("{id}/submission")]
    public async Task<IActionResult> AddSubmission(int id, [FromBody] AddSubmissionRequest req)
    {
        Console.WriteLine("=====================================================");
        Console.WriteLine($"1 Request received: {System.Text.Json.JsonSerializer.Serialize(req)}");

        try
        {
            var newSubmission = new Submission
            {
                UserQuestionId = id,
                SubmissionAnalyses = req.SubmissionAnalyses,
                CreatedAt = DateTime.UtcNow
            };

            _context.Submissions.Add(newSubmission);
            await _context.SaveChangesAsync();

            return StatusCode(201, newSubmission);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error adding submission:{ex}");
            return BadRequest(new { message = $"Error adding submission:{ex.Message}" });
        }
    }
}
